numpy/lib/function_base.py

```diff
@@ -3914,74 +3921,57 @@ def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
             "interpolation can only be 'linear', 'lower' 'higher', "
             "'midpoint', or 'nearest'")
 
-    n = np.array(False, dtype=bool)  # check for nan's flag
-    if np.issubdtype(indices.dtype, np.integer):  # take the points along axis
-        # Check if the array contains any nan's
-        if np.issubdtype(a.dtype, np.inexact):
-            indices = concatenate((indices, [-1]))
-
-        ap.partition(indices, axis=axis)
-        # ensure axis with q-th is first
-        ap = np.moveaxis(ap, axis, 0)
-        axis = 0
-
-        # Check if the array contains any nan's
-        if np.issubdtype(a.dtype, np.inexact):
-            indices = indices[:-1]
-            n = np.isnan(ap[-1:, ...])
-
-        if zerod:
-            indices = indices[0]
-        r = take(ap, indices, axis=axis, out=out)
+    # ensure axis with q-th is first, as the qth axis is first in the output
+    ap = np.moveaxis(ap, axis, 0)
+    axis = 0
 
-    else:  # weight the points above and below the indices
-        indices_below = floor(indices).astype(intp)
-        indices_above = indices_below + 1
-        indices_above[indices_above > Nx - 1] = Nx - 1
+    if np.issubdtype(indices.dtype, np.integer):
+        # take the points along axis
 
         # Check if the array contains any nan's
         if np.issubdtype(a.dtype, np.inexact):
-            indices_above = concatenate((indices_above, [-1]))
+            # nans sort to the end
+            ap.partition(concatenate((indices.ravel(), [-1])), axis=0)
+            n = np.isnan(np.take(ap, -1, axis=0))
+        else:
+            ap.partition(indices.ravel(), axis=0)
+            n = np.array(False, dtype=bool)
 
-        ap.partition(concatenate((indices_below, indices_above)), axis=axis)
+        r = take(ap, indices, axis=0, out=out)
+        if out is not None:
+            r = out  # workaround for gh-16276
 
-        # ensure axis with q-th is first
-        ap = np.moveaxis(ap, axis, 0)
-        axis = 0
+    else:
+        # weight the points above and below the indices
 
-        weights_shape = [1] * ap.ndim
-        weights_shape[axis] = len(indices)
-        weights_above = (indices - indices_below).reshape(weights_shape)
+        indices_below = not_scalar(floor(indices)).astype(intp)
+        indices_above = not_scalar(indices_below + 1)
+        indices_above[indices_above > Nx - 1] = Nx - 1
 
         # Check if the array contains any nan's
```

Comment:
Maybe combine with comment below if? and add comment to second `n = False  # Cannot contain nan`?